3p/ampanalytics-lib.js

```diff
@@ -0,0 +1,230 @@
+    this.sentinel_ = dev().assertString(
+        tryParseJson(this.win_.name, {}).sentinel,
+        'Invalid/missing sentinel on iframe name attribute' + this.win_.name);
+    if (!this.sentinel_) {
+      return;
+    }
+
+    /**
+     * Multiple creatives on a page may wish to use the same type of
+     * amp-analytics tag. This object provides a mapping between the
+     * IDs which identify which amp-analytics tag a message is to/from,
+     * with each ID's corresponding AmpAnalytics3pCreativeMessageRouter,
+     * which is an object that handles messages to/from a particular creative.
+     * @private {!Object<string, !AmpAnalytics3pCreativeMessageRouter>}
+     */
+    this.creativeMessageRouters_ = {};
+
+    /**
+     * Handles communication between frames
+     * @private {!IframeMessagingClient}
+     */
+    this.iframeMessagingClient_ = new IframeMessagingClient(win);
+    this.iframeMessagingClient_.setSentinel(this.sentinel_);
+    this.iframeMessagingClient_.registerCallback(
+        AMP_ANALYTICS_3P_MESSAGE_TYPE.CREATIVE,
+        messageContainer => {
+          let entries;
+          dev().assert(messageContainer.data &&
+              (entries = Object.entries(messageContainer.data)).length,
+              'Received empty new creative message');
+          entries.forEach(entry => {
+            const creativeId = entry[0];
+            const extraData = entry[1];
+            dev().assert(!this.creativeMessageRouters_[creativeId],
+                'Duplicate new creative message for ' + creativeId);
+            this.creativeMessageRouters_[creativeId] =
+              new AmpAnalytics3pCreativeMessageRouter(
+                this.win_, this.iframeMessagingClient_, creativeId, extraData);
+          });
+        });
+    this.iframeMessagingClient_.registerCallback(
+        AMP_ANALYTICS_3P_MESSAGE_TYPE.EVENT,
+        messageContainer => {
+          let entries;
+          dev().assert(messageContainer.data &&
+              (entries = Object.entries(messageContainer.data)).length,
+              'Received empty events message');
+          entries.forEach(entry => {
+            const creativeId = entry[0];
+            const messages = entry[1];
+            try {
+              dev().assert(messages && messages.length,
+                  'Received empty events list for' + creativeId);
+              dev().assert(this.creativeMessageRouters_[creativeId],
+                  'Discarding event message received prior to new creative' +
+                  ' message for' + creativeId);
+              this.creativeMessageRouters_[creativeId]
+                  .sendMessagesToListener(messages);
+            } catch (e) {
+              dev().error(TAG_, 'Failed to pass message to event listener: ' +
+                  e.message);
+            }
+          });
+        });
+    this.iframeMessagingClient_.sendMessage(
+        AMP_ANALYTICS_3P_MESSAGE_TYPE.READY);
+  }
+}
+
+if (window.onNewAmpAnalyticsInstance) {
```

Comment:
OK so we now require that onNewAmpAnalyticsInstance exist BEFORE adding the library?  It's not only getting called once per frame?  That can't be correct... we want it to be called for every new amp-analytics type=<vendor> instance encountered on the publisher page.  The flow should be:

- on new creative message check for onNewAmpAnalyticsInstance and if not present do not create router and warn.  Note that you will get event messages.  It's up to you how to handle them but I think its fine to just drop with warn message (perhaps worded indicated it may not have been able to be created)
- create instance of message router and pass to onNewAmpAnalyticsInstance